Sources/PromotedCore/Model/IDProducer.swift

```diff
@@ -1,37 +1,88 @@
 import Foundation
 
 /**
- Allows us to read values for session IDs before starting
+ Manages values for ancestor ID sequences.
+
+ Allows clients to read pending ancestor IDs before starting
  a session and keep those IDs consistent when the session
  does start.
+
+ Clients can also assign custom values to the ID.
  */
 final class IDProducer {
 
   typealias Producer = () -> String
-  
+
   private let initialValueProducer: Producer
   private let nextValueProducer: Producer
-  private(set) var hasAdvancedFromInitialValue: Bool
-  lazy var currentValue: String = { initialValueProducer() } ()
-  
+  private var valueHasBeenSet: Bool
+  private lazy var initialValue: String = initialValueProducer()
+
+  /// State of ID in producer, which will be used as the
+  /// ancestor ID in logged events.
+  /// If `nextValue()` has never been called, and this
+  /// hasn't been set to an external value, returns `nil`.
+  /// Set this value to assign an external ID for use as
+  /// this ancestor ID. The external ID you assign will
+  /// remain until either you set a new external value,
+  /// or you call `nextValue()` on this producer.
+  /// You can also set this value to `nil` to clear this ID
+  /// on subsequent logged events.
+  var currentValue: String? {
+    didSet { valueHasBeenSet = true }
+  }
+
+  /// Current or pending ID, which can be used to read what
+  /// the *next* autogenerated ID for `currentValue` will
+  /// be.
+  /// - If `nextValue()` has never been called, returns the
+  ///   initial value. Use this to read the pending initial
+  ///   ID before the corresponding ancestor event has been
+  ///   logged.
+  /// - Otherwise, returns `currentValue`.
+  var currentOrPendingValue: String? {
```

Comment:
What's the issue with immediately setting the currentValue to the initialValue?

Reply:
Because `currentValue` will be logged to events as the ancestor id. If I immediately set it, then the ancestor id will be included before they are specified.

Reply:
I simplified this code (eliminated 2 internal fields), perhaps the new version is more clear.